### PR TITLE
Fix remove unsafe path joins

### PR DESCRIPTION
**Pull request by fruttasecca**
Related: https://github.com/orchest/orchest/pull/907

Fixes: https://github.com/orchest/orchest/issues/906

Credits to @porcupineyhairs for discovery and fix proposal

There are a couple of regressions on `dev` that make it not possible to test all edge cases, it's something we will have to look into next week.

### CODE REVIEW

File: services/orchest-webserver/app/app/core/projects.py [hunk 1]
```diff
@@ -94,7 +92,7 @@ def _collateral(self, project_uuid: str, project_path: str):
         # `.orchest/` directory because an existing project might be
         # added to Orchest and already contain a root-level
         # `.gitignore`, which we don't want to inject ourselves in.
-        expected_git_ignore_file = os.path.join(
+        expected_git_ignore_file = safe_join(
```
Comment (yannickperrenet) [inline services/orchest-webserver/app/app/core/projects.py:95]: Have you tested this? 

Given that Flask's `safe_join` only takes two arguments and in later versions (see https://github.com/pallets/flask/commit/b7501776a12c15e64b6cdac1e3b903a00df7d285) uses Werkzeug's `safe_join` that does take multiple arguments, and us running `Flask=1.1.2`, I don't think this will actually work.

File: services/orchest-webserver/app/app/core/projects.py [hunk 2]
```diff
@@ -107,7 +105,7 @@ def _collateral(self, project_uuid: str, project_path: str):
 
         # Initialize with default environments only if the project has
         # no environments directory.
-        expected_env_dir = os.path.join(full_project_path, ".orchest", "environments")
+        expected_env_dir = safe_join(full_project_path, ".orchest", "environments")
```
Comment (yannickperrenet) [inline services/orchest-webserver/app/app/core/projects.py:108]: Here are also more than 2 arguments, so need to verify. As per https://github.com/orchest/orchest/pull/913/files#r861897137

File: services/orchest-webserver/app/app/core/projects.py (hunk 1 above)
Comment (fruttasecca) [reply services/orchest-webserver/app/app/core/projects.py:95]: Yes I have tested every change AFAIK, I've also just rebuilt the webserver image and run the following in a python repl inside it
```
>>> import flask
>>> flask.__version__
'1.1.2'
>>> from flask import safe_join
>>> safe_join("1", "2", "3")
'1/2/3'
>>>
```

**Review by yannickperrenet — APPROVED** [review]
:100: 

Thank you @porcupineyhairs!

File: services/orchest-webserver/app/app/utils.py [hunk 3]
```diff
@@ -629,7 +623,7 @@ def pipeline_set_notebook_kernels(pipeline_json, pipeline_directory, project_uui
 
         if "ipynb" == step["file_path"].split(".")[-1]:
 
-            notebook_path = os.path.join(pipeline_directory, step["file_path"])
+            notebook_path = safe_join(pipeline_directory, step["file_path"])
```
Comment (ricklamers) [inline services/orchest-webserver/app/app/utils.py:627]: I expect this to cause problems because file paths can be relative and point to files above the pipeline directory. We should find all similar resolves of paths that use the pipeline step file path in particular.

Comment (fruttasecca) [reply services/orchest-webserver/app/app/utils.py:627]: Good catch, will take a look

File: services/orchest-webserver/app/app/utils.py [hunk 4]
```diff
@@ -749,6 +751,23 @@ def is_valid_project_relative_path(project_uuid, path: str) -> str:
...
+    )
+    pipeline_path = get_pipeline_path(pipeline_uuid, project_uuid)
+    abs_path = os.path.abspath(
+        os.path.normpath(os.path.join(os.path.split(pipeline_path)[0], path))
+    )
+    return abs_path.startswith(project_path)
+
+
+def is_valid_data_path(path: str):
+    return os.path.abspath(os.path.normpath(path)).startswith("/data")
```
Comment (iannbing) [inline services/orchest-webserver/app/app/utils.py:768]: @fruttasecca  Isn't this line always returning `false`? `os.path.abspath(os.path.normpath(path))` most likely would start with `/userdir`?

Comment (iannbing) [reply services/orchest-webserver/app/app/utils.py:768]: I propose changing it to https://github.com/orchest/orchest/blob/6f66315c05b8b9a3c2061c59ab7279c4510125c0/services/orchest-webserver/app/app/utils.py#L768 in PR #914. Please review the change and discuss in the said PR. :-)
